File: dipeo/domain/service_utils.py
"""
Utility functions for working with service types
These utilities help distinguish between LLM and non-LLM services
"""


from dipeo.diagram_generated import APIServiceType, LLMService

# Set of APIServiceType values that are LLM services
LLM_SERVICE_TYPES: set[APIServiceType] = {
    APIServiceType.OPENAI,
    APIServiceType.ANTHROPIC,
    APIServiceType.GOOGLE,
    APIServiceType.GEMINI,
    APIServiceType.BEDROCK,
    APIServiceType.VERTEX,
    APIServiceType.DEEPSEEK,
}


def is_llm_service(service: APIServiceType) -> bool:
    """Check if an APIServiceType is an LLM service"""
    return service in LLM_SERVICE_TYPES


def api_service_type_to_llm_service(service: APIServiceType) -> LLMService:
    """
    Convert APIServiceType to LLMService
    
    Raises:
        ValueError: If the APIServiceType is not an LLM service
    """
    if not is_llm_service(service):
        raise ValueError(f'APIServiceType "{service.value}" is not an LLM service')
    
    # Handle special cases
    if service == APIServiceType.GEMINI:
        return LLMService.GOOGLE
    
    # Direct mapping for others
    return LLMService(service.value)


def llm_service_to_api_service_type(service: LLMService) -> APIServiceType:
    """Convert LLMService to APIServiceType"""
    # Direct mapping works for all LLMService values
    return APIServiceType(service.value)


def get_llm_service_types() -> list[APIServiceType]:
    """Get all LLM service types"""
    return list(LLM_SERVICE_TYPES)


def get_non_llm_service_types() -> list[APIServiceType]:
    """Get all non-LLM service types"""
    return [service for service in APIServiceType if service not in LLM_SERVICE_TYPES]


def is_valid_llm_service(service: str) -> bool:
    """Type guard to check if a string is a valid LLMService"""
    try:
        LLMService(service)
        return True
    except ValueError:
        return False


def is_valid_api_service_type(service: str) -> bool:
    """Type guard to check if a string is a valid APIServiceType"""
    try:
        APIServiceType(service)
        return True
    except ValueError:
        return False

File: dipeo/models/src/service-utils.ts
/**
 * Utility functions for working with service types
 * These utilities help distinguish between LLM and non-LLM services
 */

import { LLMService, APIServiceType } from './integration.js';

/**
 * Set of APIServiceType values that are LLM services
 */
const LLM_SERVICE_TYPES = new Set<APIServiceType>([
  APIServiceType.OPENAI,
  APIServiceType.ANTHROPIC,
  APIServiceType.GOOGLE,
  APIServiceType.GEMINI,
  APIServiceType.BEDROCK,
  APIServiceType.VERTEX,
  APIServiceType.DEEPSEEK,
]);

/**
 * Check if an APIServiceType is an LLM service
 */
export function isLLMService(service: APIServiceType): boolean {
  return LLM_SERVICE_TYPES.has(service);
}

/**
 * Convert APIServiceType to LLMService
 * @throws Error if the APIServiceType is not an LLM service
 */
export function apiServiceTypeToLLMService(service: APIServiceType): LLMService {
  if (!isLLMService(service)) {
    throw new Error(`APIServiceType "${service}" is not an LLM service`);
  }
  
  // Handle special cases
  if (service === APIServiceType.GEMINI) {
    return LLMService.GOOGLE;
  }
  
  // Direct mapping for others
  return service as unknown as LLMService;
}

/**
 * Convert LLMService to APIServiceType
 */
export function llmServiceToAPIServiceType(service: LLMService): APIServiceType {
  // Direct mapping works for all LLMService values
  return service as unknown as APIServiceType;
}

/**
 * Get all LLM service types
 */
export function getLLMServiceTypes(): APIServiceType[] {
  return Array.from(LLM_SERVICE_TYPES);
}

/**
 * Get all non-LLM service types
 */
export function getNonLLMServiceTypes(): APIServiceType[] {
  return Object.values(APIServiceType).filter(service => !isLLMService(service));
}

/**
 * Type guard to check if a string is a valid LLMService
 */
export function isValidLLMService(service: string): service is LLMService {
  return Object.values(LLMService).includes(service as LLMService);
}

/**
 * Type guard to check if a string is a valid APIServiceType
 */
export function isValidAPIServiceType(service: string): service is APIServiceType {
  return Object.values(APIServiceType).includes(service as APIServiceType);
}